File: 0401-0500/0456-132-pattern.rs
// TODO: 还有更简单的解法，不需要mins
struct Solution {}

impl Solution {
    pub fn find132pattern(nums: Vec<i32>) -> bool {
        if nums.is_empty() {
            return false;
        }
        let mut mins = Vec::new();
        let mut mi = nums[0];
        for &x in nums.iter() {
            mi = mi.min(x);
            mins.push(mi);
        }
        let mut st = Vec::new();  // 最小栈
        // println!("data: {:?}", nums);
        // println!("mins: {:?}", mins);
        // 注意j开始的位置
        for j in (1..nums.len()).rev() {
            // 下面的条件也可以不加
            if nums[j] > mins[j] {
                while !st.is_empty() && *st.last().unwrap() <= mins[j] {
                    st.pop();
                }
                // 如果st有元素，刚好比mins[j]大
                if !st.is_empty() && *st.last().unwrap() < nums[j] {
                    return true;
                }
                // 栈顶 >= nums[j]
                st.push(nums[j]);
                // println!("st: {:?}", st);
            }
        }
        false
    }
}

fn main() {
    let nums = vec![1, 2, 3, 4];
    assert_eq!(Solution::find132pattern(nums), false);

    let nums = vec![3, 1, 4, 2];
    assert_eq!(Solution::find132pattern(nums), true);

    let nums = vec![3, 5, 0, 3, 4];
    assert_eq!(Solution::find132pattern(nums), true);
    /*
     * nums: 3, 5, 0, 3, 4
     * mins: 3, 3, 0, 0, 0
     */
}

File: 0401-0500/0456-132-pattern.cpp
#include "common.h"

#define SOLUTION 2

#if SOLUTION == 1
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
      const int n = nums.size();
      if (n < 3) return false;
      /*
       * 给你一个整数数组 nums ，数组中共有 n 个整数。132 模式的子序列 由三个整数 nums[i]、nums[j] 和 nums[k] 组成，并同时满足：i < j < k 和 nums[i] < nums[k] < nums[j]
       */
      multiset<int> rights;
      int left_min = nums[0];
      for (int i = 2; i < n; ++i) {
        rights.insert(nums[i]);
      }
      for (int i = 1; i < n - 1; ++i) {
        if (nums[i] > left_min) {
          auto p = rights.upper_bound(left_min);
          if (p != rights.end() && *p < nums[i]) {
            return true;
          }
        }
        left_min = min(left_min, nums[i]);
        // 需要传迭代器给erase函数，否则会删除等于这个值的所有元素
        rights.erase(rights.find(nums[i + 1]));
      }
      return false;
    }
};
#elif SOLUTION == 2

// TODO: 单调栈一次遍历的方法
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
      // 132
    }
};

#endif

int main() {
  vector<int> nums;
  INPUT_ARRAY(nums);
  cout << Solution().find132pattern(nums) << endl;
  return 0;
}
